usr/src/uts/common/fs/zfs/dbuf.c

```diff
@@ -3284,13 +3284,14 @@ dbuf_write_children_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
 	dmu_buf_impl_t *db = vdb;
 	dnode_t *dn;
 	blkptr_t *bp;
-	uint64_t i;
+	int i;
```

Comment:
You're right that this could overflow in that case.  I'd suggest fixing this by changing the assertion that's a few lines down, to assert that epbs is < 31.  (in practice it can be no more than 17: `SPA_MAXBLOCKSHIFT - SPA_BLKPTRSHIFT`)

Are you thinking that lint should complain about any comparison of a signed number with the result of a left-shift?